src/chips/nrf51822/gpio.rs

```diff
@@ -0,0 +1,121 @@
+use core::mem;
+use core::ops::{Index, IndexMut};
+use hil;
+
+// Source: https://github.com/hackndev/zinc/tree/master/volatile_cell
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct VolatileCell<T> {
+    value: T,
+}
+
+#[allow(dead_code)]
+impl<T> VolatileCell<T> {
+    #[inline]
+    pub fn get(&self) -> T {
+        unsafe {
+            ::core::intrinsics::volatile_load(&self.value)
+        }
+    }
+
+    #[inline]
+    pub fn set(&self, value: T) {
+        unsafe {
+            ::core::intrinsics::volatile_store(&self.value as *const T as *mut T, value)
+        }
+    }
+}
+
+#[allow(non_snake_case)]
+struct GPIO {
+    _pad0: [u8; 1284],
+    pub OUT: VolatileCell<u32>,
+    pub OUTSET: VolatileCell<u32>,
+    pub OUTCLR: VolatileCell<u32>,
+    pub IN: VolatileCell<u32>,
+    pub DIR: VolatileCell<u32>,
+    pub DIRSET: VolatileCell<u32>,
+    pub DIRCLR: VolatileCell<u32>,
+    _pad1: [u8; 480],
+    pub PIN_CNF: [VolatileCell<u32>; 32],
+}
+
+#[allow(non_snake_case)]
+fn GPIO() -> &'static GPIO {
+    unsafe { mem::transmute(0x50000000 as usize) }
+}
+
+pub struct GPIOPin {
+    pin: u8,
+}
+
+impl GPIOPin {
+    const fn new(pin: u8) -> GPIOPin {
+        GPIOPin { pin: pin }
+    }
+}
+
+impl hil::gpio::GPIOPin for GPIOPin {
+    fn enable_output(&self) {
+        GPIO().PIN_CNF[self.pin as usize].set((1 << 0) | (1 << 1) | (0 << 2) | (0 << 8) | (0 << 16));
+    }
+
+    fn enable_input(&self, _mode: hil::gpio::InputMode) {
+        unimplemented!();
+    }
+
+    fn disable(&self) {
+        unimplemented!();
+    }
+
+    fn set(&self) {
+        GPIO().OUTSET.set(1 << self.pin);
+    }
+
+    fn clear(&self) {
+        GPIO().OUTCLR.set(1 << self.pin);
+    }
+
+    fn toggle(&self) {
+        unimplemented!();
+    }
+
+    fn read(&self) -> bool {
+        unimplemented!();
+    }
+
+    fn enable_interrupt(&self, _identifier: usize, _mode: hil::gpio::InterruptMode) {
+        unimplemented!();
+    }
+}
+
+pub struct Port {
```

Comment:
I think the nrf51822 has only one port, so the abstraction into a `Port` may not end up being very useful.